### DISCUSSION TITLE
A couple of basic problems

### DISCUSSION THREAD

**Discussion by lymanc** [lamemakes/pilot-drive#20] [Troubleshooting]
First some background. My interest is to have Android Auto like capabilities in my classic 1971 BMW Bavaria without molesting the dash or center console.  To this end I have found a way to mount a thin tablet such as the Google Pixel Slate, and if additional electronics are needed such a computer, router, or whatever they would be mounted under the rear seat. I am more of a user than developer as my development skills a quite dated.

At this pointed I have installed pilot-drive on Ubuntu 22.04, rpi 4 and a google pixel slate. On pixel slate it runs under "Debian GNU/Linux trixie/sid" which is run as a virtual container under "re-imagined" Chrome OS. 

First the simpler one to explain is with Pixel Slate. The command "sudo python3.11 -m pip install pilot-drive" produces "error: external-managed-environment" followed by a boilerplate explanation which I don't fully understand. It part it suggest creating a virtual environment, which I did with the same result.

In the case of ubuntu and rpi, they both installed pilot-drive and initially ran pilot_drive successfully. But after i ran pilot-drive --setup and selected 'Android' under "Use a phone with PILOT Drive?", starting pilot_drive produced:
 
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/usr/local/lib/python3.11/site-packages/pilot_drive/__main__.py", line 74, in <module>
    start()
  File "/usr/local/lib/python3.11/site-packages/pilot_drive/__main__.py", line 47, in start
    run()
  File "/usr/local/lib/python3.11/site-packages/pilot_drive/__main__.py", line 59, in run
    pilot_drive = PilotDrive()
                  ^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/pilot_drive/pd_manager.py", line 80, in __init__
    self.phone: Phone = self.service_factory(
                        ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/pilot_drive/pd_manager.py", line 121, in service_factory
    new_service = service(
                  ^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/pilot_drive/services/phone/phone.py", line 77, in __init__
    self.__phone_manager = AndroidManager(logger=self.logger)
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/pilot_drive/services/phone/android_manager.py", line 69, in __init__
    self.__validate_dependencies()  # Confirm all dependencies are there
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/pilot_drive/services/phone/android_manager.py", line 103, in __validate_dependencies
    raise AdbDependenciesMissingException(
pilot_drive.services.phone.android_manager.AdbDependenciesMissingException: Missing dependency: AAPT2, bash output returned: /bin/sh: 1: aapt2: Permission denied
 
In both cases, removing pilot-drive, and reinstalling it does not overcome the problem.  

Hopefully I have clearly stated my findings. 

John

**Comment by lamemakes** [comment]
Hey @lymanc! Thanks for the follow up here.

> 
> On pixel slate it runs under "Debian GNU/Linux trixie/sid" which is run as a virtual container under "re-imagined" Chrome OS.
> 
> First the simpler one to explain is with Pixel Slate. The command "sudo python3.11 -m pip install pilot-drive" produces "error: external-managed-environment" followed by a boilerplate explanation which I don't fully understand. It part it suggest creating a virtual environment, which I did with the same result.

This is super interesting and I'd love to potentially coordinate more to work through some of the slate kinks. I primarily developed & tested PILOT Drive on Ubuntu, Fedora & Raspberry Pi OS but would really enjoy to expand support for other platforms that I might not have immediate access to like that. A quick [stackoverflow search](https://stackoverflow.com/a/75722775) suggested tinkering with pip a little if venv was giving you issues.

> In the case of ubuntu and rpi, they both installed pilot-drive and initially ran pilot_drive successfully. But after i ran pilot-drive --setup and selected 'Android' under "Use a phone with PILOT Drive?", starting pilot_drive produced:

For this, try running the setup as `sudo`. Due to the interactions that PILOT Drive needs to have with some of the Linux communication buses, it needs to be installed, setup, and run with root privileges. So try `sudo pilot-drive --setup` and go through the prompts to see if your results are any different.

Thanks for your patience and giving PILOT Drive a go! Feedback like this helps improve my documentation and the overall software a ton. Cheers!

**Comment by lymanc** [comment]
Pixel Slate is what I utimately want to use as my AA interface. BTW, used prices for Pixel Slates on ebay are approaching the cost of a comparaible rpi based setup.

With Pixel slate, it didn't make any difference whether I tried to start PILOT in a virtual environment or not.

Regarding Ubuntu and rpi, followed your expample and have always used sudo with and without setup. Just to be sure I just went throught it a gain using sudo, and no cigar.

Regarding documentation, no mention is make whether or not to use a virtual environment, or whether it makes a differnce, etc. I got a feeling that doing the manual install without a virtual environment may invoke some warnings. 

Given my lack of experience with python, I am dead in the water for now.

Thank you for your prompt reply.

**Comment by lymanc** [comment]
See  if this helps:

Step 1) Fresh install and start following the "Quick Start" procedure.
	 -- things look --
Step 2) Shutdown

Step 3) Restart PILOT with --setup argument
	-- looks good --
	
Step 4) Select Android phone and shutdown

Step 5) Restart PILOT
       -- notice some additional install activity including something about AAPT2
       -- PILOT fails to start

Step 6) Restart PILOT
	-- fails to start leaving the error message I previously posted
	 
Step 7) Restart PILOT with --setup argument, and not select phone this time.

Step 8) Restart PILOT again fails to leaving the error message.

Note: Error message makes several references related to "phone", and mentions AAPT2 even though phone was not selected this time. Suggesting the 2nd setup did not restore the system state that existed in Step 3 above.

**Comment by lymanc** [comment]
Found the problem, or at least the cause of the problem.

After nosing around a bit I discovered the file /etc/pilot-drive/config/settings.json, and found that enable 'phone' was set 'true', even after it had been set false via --setup.  Changing enable phone to false restored normal operation.  

So now I can move on under Ubuntu and RPI as  long as I don't enable 'phone'.

**Comment by lamemakes** [comment]
Thanks for the info here! This was all super helpful. A new issue has just been written up here for the setup issues: #21, if you wanted to confirm that the behavior you're seeing.

As for the AAPT2 issues, what happens when you execute `aapt2` and `sudo aapt2` in terminal?

**Comment by lamemakes** [comment]
> Regarding documentation, no mention is make whether or not to use a virtual environment, or whether it makes a differnce, etc. I got a feeling that doing the manual install without a virtual environment may invoke some warnings.

I'll also be adding a better implementation for a virtual environment. Currently I've been handling things in a way that isn't the greatest by just installing straight to the user environment but venv is the way to go - will write an issue for that too. Again, this feedback is all ace and I really appreciate it!

**Comment by lymanc** [comment]
BTW, same problem when camera is enable via --setup, and a camera is not
present.  Furthermore, disabling the camera via --setup does not set it
false in /etc/pilot_drive/config.

**Comment by lamemakes** [comment]
Yeah the issue you found is for all features in the setup tool, not just Android so that was a good catch. 

Were you able to try this?:

> As for the AAPT2 issues, what happens when you execute `aapt2` and `sudo aapt2` in terminal?

**Comment by lymanc** [comment]
locate aapt2 yield:

opt/Projects/Pilot/pilot-drive/bin/aapt2.save
/opt/Projects/Pilot/pilot-drive/bin/aapt2.save/arm64-v8a
/opt/Projects/Pilot/pilot-drive/bin/aapt2.save/armeabi-v7a
/opt/Projects/Pilot/pilot-drive/bin/aapt2.save/x86
/opt/Projects/Pilot/pilot-drive/bin/aapt2.save/x86_64
/opt/Projects/Pilot/pilot-drive/bin/aapt2.save/arm64-v8a/aapt2
/opt/Projects/Pilot/pilot-drive/bin/aapt2.save/armeabi-v7a/aapt2
/opt/Projects/Pilot/pilot-drive/bin/aapt2.save/x86/aapt2
/opt/Projects/Pilot/pilot-drive/bin/aapt2.save/x86_64/aapt2
/usr/local/bin/aapt2

/usr/local/bin/aapt2: ELF 64-bit LSB executable, x86-64, version 1 (SYSV),
statically linked, stripped

***@***.***:/opt/Projects/Pilot$ appt2
Traceback (most recent call last):
  File "/usr/lib/command-not-found", line 27, in <module>
    from CommandNotFound.util import crash_guard
ModuleNotFoundError: No module named 'CommandNotFound'

***@***.***:/opt/Projects/Pilot$ sudo appt2
[sudo] password for johnc:
sudo: appt2: command not found

**Comment by lymanc** [comment]
I forgot to check the aapt2 permissions which were rw-r--r--.  Why the
non-root execution produced an output is a mystery to me, but  does
explain why the as root execution produced 'command not found'.

**Comment by lamemakes** [comment]
It looks like you might've executed `appt2` rather than `aapt2`:

> ***@***.***:/opt/Projects/Pilot$ sudo appt2
> [sudo] password for johnc:
> sudo: appt2: command not found

That's something I've done a thousand times over myself if that's the case.

**Comment by lamemakes** [comment]
@lymanc any updates here? did any of that previously mentioned aapt2 stuff work? I just fixed the issues you identified with the installer, thanks again for that!

**Comment by lymanc** [comment]
Re: aapt2:  It had to do with user permissions.  If my memory serves me
correctly, when I initially installed Pilot I got an error message related
to the execution of aapt2.  Later  I found out that running Pilot as root
(sudo), it loaded just fine.  Changed the permissions to  include me and it
load ok as well.  Wish you the best of luck, and look forward to your
continued development of Pilot.  In the meantime, I found a configuration
which allows me to run Android Auto, on my android tablets, including
Amazon Fire HD 10, Nexus 7 -- wirelessly.     Wish you the best of luck,
and encourage your continued development of Pilot.
